Local message history

BoneFiend opened the pull request:
## Description

Implemented an MVP local overlay message history in camera system. Formatting should work fine on mobile. Limitations include losing everything when changing pages.

## Screenshots

Before a message is sent:
![image](https://user-images.githubusercontent.com/24929509/206382401-76368563-1a6d-477e-a36a-6e5905d4b76a.png)

After messages have been sent:
![image](https://user-images.githubusercontent.com/24929509/206382521-85ef4406-e31b-47c8-af7e-e2d74f7ddfdb.png)


## Steps to Test

1. Open a terminal in dashboard/client
2. Run `yarn install` then `yarn start`
3. In the web page move to Camera System tab then type some messages under Overlay Message

BoneFiend committed:
Local message history

Implemented an MVP local overlay message history in camera system. Formatting should work fine on mobile. Limitations include losing everything when changing pages.

kdav108 commented on client/src/views/common/CameraSystemView.tsx:
```diff
@@ -13,7 +13,7 @@ import OverlaySelectionContainer from 'components/common/camera_settings/Overlay
 export default function CameraSystemView() {
   return (
-    <ContentPage title="Camera Settings">
+    <ContentPage title="Camera System">
```
If we update the page title, let's update this function's documentation as well:
https://github.com/monash-human-power/dashboard/blob/a3c44a5d2a54a00e431fa9520653f829336bd59d/client/src/views/common/CameraSystemView.tsx#LL10

BoneFiend replied:
It seemed that everything referred to "Camera System" instead of "Camera Settings". It also makes more sense as there are almost no settings to change on that page.

kdav108 replied:
yeah good spotting

kdav108 commented on client/src/components/common/camera_settings/OverlayMessage.tsx:
```diff
@@ -31,6 +46,37 @@ export default function OverlayMessage({ sendMessage }: OverlayMessageProps) {
+        time /= 60 * 60 * 24;
+      }
+      unit += time >= 2 ? 's' : '';
+      newMessageHistory[m].elapsed = Math.floor(time).toString().concat(unit); // Updates dummy variable
```
> Updates dummy variable

Didn't quite get what this is referring to?

kdav108 commented on client/src/components/common/camera_settings/OverlayMessage.tsx:
```diff
@@ -31,6 +46,37 @@ export default function OverlayMessage({ sendMessage }: OverlayMessageProps) {
+
+  useEffect(() => {
+    // Updates elapsed times in messageHistory every second
+    const interval = setInterval(updateElapsed, 1000);
```
Seeing the elapsed time update every second on the UI looks really good. But we should just note in case we have issues in the future that updating every second could have performance impacts.

If we want, we could also look at updating every minute, if we don't require precision to be correct to the second. Something we could discuss with Charles and others who would use this feature.

BoneFiend replied:
This is a good point. From memory, the message pops up for 10 seconds in the bike, which means knowing whether a message is +/- 10 seconds important (as not to interrupt a previous message).

I will try to catch Charles for an interview to see what's important to him. Updating every minute after the first 60 seconds could be a good solution.

kdav108 replied:
That's a good point. Updating every minute after first 60 seconds is a possible alternative, but might be a bit harder to implement. But yeah we should get some opinions from users.

kdav108 reviewed (commented):
This is great stuff! Really impressive given it's your first attempt at React and Front-end.

The only major discussion I think we need to have is how we can make the message history persistent (i.e. stays the same even when we refresh or change between pages). An easy fix is to store the message history in the backend (`/server/`) just like we do for some other data: https://github.com/monash-human-power/dashboard/blob/a3c44a5d2a54a00e431fa9520653f829336bd59d/server/js/sockets.js#L23-L37

At the same time, I feel like there may be value in knowing the message history regardless of which user has sent in. The current method and the suggested way to retain above will only show messages sent from the same user, say A, who's running this on the browser, but if another user, B, on another device sends something, A and B cannot see each other's messages. If we want everyone to see the messages, this becomes an architectural decision that spans more than just this repo. We need to figure out which of our software systems (and maybe we have to build a new one) stores all the message history centrally and then communicates with everyone.

BoneFiend committed:
Refactored code + added row borders

Refactored message history table component to be more inline with React standard, after reading more docs. Also added nicknames for V3 (Bilby),

BoneFiend committed:
Added local storage

Now using localStorage so locally sent messages are kept in between sessions. Included footer button to delete messages

kdav108 reviewed (commented):
This is working great! Lots of really good React code here :)

Just a comment on the architecture, we're using local storage on the browser here. All our other approaches of persisting data (like the WM and camera status, and the speed vs distance graph) use a different strategy (they save the data in a [local variable](https://github.com/monash-human-power/dashboard/blob/48528324d45b6d0325f7f48f144eb3c31a411699/server/js/sockets.js#L23-L37) in the server and [read it from there upon refresh](https://github.com/monash-human-power/dashboard/blob/48528324d45b6d0325f7f48f144eb3c31a411699/client/src/components/v3/StatisticRow.tsx#L24-L26)). Just FYI, both method still work!

BoneFiend committed:
Small efficiency gains + bug squashing

BoneFiend commented:
Much cleaner interface now with only important information in large writing.

![image](https://user-images.githubusercontent.com/24929509/227186715-393a919d-6478-41bb-b2fa-eef097d9b8d3.png)

The live elapsed time is now more efficient and will update less after a minute.

All messages are stored locally on device which sent them. Can also be deleted easily.

Can now toggle between 12 hour and 24 hour time.

AroshHeenkenda reviewed (approved):
I approve this code.
